src/CentrifugeRouter.sol

```diff
@@ -2,97 +2,197 @@
-        uint256 lockedRequest = lockedRequests[user][vault];
-        require(lockedRequest > 0, "CentrifugeRouter/user-has-no-balance");
-        lockedRequests[user][vault] = 0;
-        IERC7540Vault(vault).requestDeposit(lockedRequest, user, address(this));
-        emit ExecuteLockedDepositRequest(vault, user);
+    function executeLockedDepositRequest(address vault, address controller) external protected {
+        uint256 lockedRequest = lockedRequests[controller][vault];
+        require(lockedRequest > 0, "CentrifugeRouter/controller-has-no-balance");
+        lockedRequests[controller][vault] = 0;
+
+        address asset = poolManager.vaultToAsset(vault);
+
+        require(asset != address(0), "CentrifugeRouter/unknown-vault");
+        EscrowLike(escrow).approve(asset, address(this), lockedRequest);
+        SafeTransferLib.safeTransferFrom(asset, escrow, address(this), lockedRequest);
+
+        _approveMax(asset, vault);
+        IERC7540Vault(vault).requestDeposit(lockedRequest, controller, address(this));
+        emit ExecuteLockedDepositRequest(vault, controller, _initiator);
     }
 
     /// @inheritdoc ICentrifugeRouter
-    function claimDeposit(address vault, address user) external {
-        uint256 maxDeposit = IERC7540Vault(vault).maxDeposit(user);
-        IERC7540Vault(vault).deposit(maxDeposit, user, user);
+    function claimDeposit(address vault, address receiver, address controller) external protected {
+        uint256 maxDeposit = IERC7540Vault(vault).maxDeposit(controller);
+        IERC7540Vault(vault).deposit(maxDeposit, receiver, controller);
     }
 
     // --- Redeem ---
     /// @inheritdoc ICentrifugeRouter
-    function requestRedeem(address vault, uint256 amount) external {
-        IERC7540Vault(vault).requestRedeem(amount, msg.sender, msg.sender);
+    function requestRedeem(address vault, uint256 amount, address controller, address owner) external protected {
+        IERC7540Vault(vault).requestRedeem(amount, controller, owner);
+    }
+
+    /// @inheritdoc ICentrifugeRouter
+    function claimRedeem(address vault, address receiver, address controller) external protected {
+        uint256 maxRedeem = IERC7540Vault(vault).maxRedeem(controller);
+        IERC7540Vault(vault).redeem(maxRedeem, receiver, controller);
     }
 
+    // --- ERC20 permits ---
     /// @inheritdoc ICentrifugeRouter
-    function claimRedeem(address vault, address user) external {
-        uint256 maxRedeem = IERC7540Vault(vault).maxRedeem(user);
-        IERC7540Vault(vault).redeem(maxRedeem, user, user);
+    function permit(address asset, address spender, uint256 assets, uint256 deadline, uint8 v, bytes32 r, bytes32 s)
+        external
+        protected
+    {
+        try IERC20Permit(asset).permit(_initiator, spender, assets, deadline, v, r, s) {} catch {}
+    }
+
+    // --- ERC20 wrapping ---
+    function wrap(address wrapper, uint256 amount, address receiver) external protected {
```

Comment:
let's also include the following check here: 
`require(receiver != address(0), "CentrifugeRouter/zero-address");`
to be consistent with `unwrap`

Reply:
I removed it from unwrap now, it seems like a useless check to me.